(CDAP-16353) Move get artifact details for range into ArtifactRepositoryReader

wyzhang opened the pull request:
Motivation:
AppLifecycleDeployment use getArtifactDetails to find the artifact detail
for deploymnet. Moving this method into ArtifactRepositoryReader allows us
to have remote vs local implication used by preview (when configured with
local levelDB) and appfabric respectively.

Add implementation for this method in both LocalArtifactRepositoryReader
and RemoteArtifactRepositoryReader. But the latter is not being used
anywhere at the moment. A follow-up change will make preview to use
RemoteArtifactRepositoryReader when configured with local levelDB.

albertshau commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java:
```diff
@@ -453,12 +454,15 @@ private boolean loadSystemArtifacts(ExecutorService executorService,
     // delete the artifact first and then privileges. Not the other way to avoid orphan artifact
     // which does not have any privilege if the artifact delete from store fails. see CDAP-6648
+    LOG.debug("wyzhang: delete artiact store: " + artifactStore);
+    LOG.debug("wyzhang: delete artiact id: " + artifactId.toString());
```
remove debug logs

wyzhang replied:
Done.

Forgot to clean them up.

albertshau commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepositoryReader.java:
```diff
@@ -87,6 +92,30 @@ public ArtifactDetail getArtifact(Id.Artifact artifactId) throws Exception {
+                               range.getLower().toString(),
+                               range.getUpper().toString(),
+                               String.valueOf(limit),
+                               String.valueOf(order));
```
order.name()

wyzhang replied:
Done

albertshau commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepositoryReader.java:
```diff
@@ -87,6 +92,30 @@ public ArtifactDetail getArtifact(Id.Artifact artifactId) throws Exception {
+    String url = String.format("namespaces/%s/artifacts/%s/versions?lower=%s&upper=%s&limit=%s&order=%s",
+                               range.getNamespace(),
+                               range.getName(),
+                               range.getLower().toString(),
```
I assume this means both lower and upper are always non-null?

wyzhang replied:
Yes. 

Version fields in ArtifactRange are not annotated with nullable. 
Also quickly checked a number of callers, I don't see any null being used to construct ArtifactRange.

albertshau commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepositoryReader.java:
```diff
@@ -87,6 +92,30 @@ public ArtifactDetail getArtifact(Id.Artifact artifactId) throws Exception {
+    List<ArtifactDetail> detailList = new ArrayList<>();
+    for (ArtifactDetail detail : details) {
+      Location artifactLocation =
+        Locations.getLocationFromAbsolutePath(locationFactory, detail.getDescriptor().getLocationURI().getPath());
```
is there a reason this utility method is being used over locationFactory.create()? Are they different?

wyzhang replied:
Fixed. We can use LocationFactory.create() directly. The util cleans up and validate the path. But we are guaranteed the location URI is valid, so can just use create.

wyzhang committed:
(CDAP-16353) Move get artifact details for range into ArtifactRepositoryReader

Motivation:
AppLifecycleDeployment use getArtifactDetails to find the artifact detail
for deploymnet. Moving this method into ArtifactRepositoryReader allows us
to have remote vs local implication used by preview (when configured with
local levelDB) and appfabric respectively.

Add implementation for this method in both LocalArtifactRepositoryReader
and RemoteArtifactRepositoryReader. But the latter is not being used
anywhere at the moment. A follow-up change will make preview to use
RemoteArtifactRepositoryReader when configured with local levelDB.
